algebird-core/src/main/scala/com/twitter/algebird/Windows.scala

```diff
@@ -1,142 +1,99 @@
-      }
+      val i = stepsFrom(time, currTime)
+      require(i >= 1, s"currTime=$currTime, step=$step, time=$time, i=$i")
+
+      // moving forward by i (>= 1) time steps
+      val w = Windows(buf.step(i), step, time + (i * step))
       require(currTime < w.time + w.step, s"currTime=$currTime, time=$time, w.time=${w.time}")
       require(currTime >= w.minTime)
       w
     }
 
-  def minTime: Long =
-    time - (slots.size - 1) * step
-
-  private[this] def slotContains(i: Int, currTime: Long): Boolean = {
-    // i=4, index=5, indexDelta=1
-    // i=6, index=5, indexDelta=slots.size - 1
-    val n = slots.size
-    val indexDelta = ((index - i) + n) % n
-    val start = time - indexDelta * step
-    start <= currTime && currTime < (start + step)
-  }
-
-  // invariant: currTime must in [minTime, time + step).
-  private[this] def getSlot(currTime: Long): Int = {
-    require(minTime <= currTime)
-    require(currTime < time + step)
-    @tailrec def loop(thisIndex: Int): Int =
-      if (slotContains(thisIndex, currTime)) thisIndex
-      else loop(makeIndex(thisIndex - 1))
-    loop(index)
-  }
-
+  /**
+   * Adding something before minTime is a no-op.
+   */
   def add(a: A, currTime: Long)(implicit ev: Monoid[A]): Windows[A] =
     if (currTime < minTime) this
     else if (currTime < time + step) {
-      val i = getSlot(currTime)
-      val sum = ev.plus(slots(i), a)
-      copy(slots = slots.updated(i, sum))
+      copy(buf = buf.add(a, stepsFrom(currTime, time)))
     } else step(currTime).add(a, currTime)
 
-  def lowerBoundSum(implicit ev: Monoid[A]): A = {
-    val before = ev.sum(slots.iterator.take(index - 1)) // before
-    val after = ev.sum(slots.iterator.drop(index)) // after
-    ev.plus(before, after)
-  }
-
-  def upperBoundSum(implicit ev: Monoid[A]): A =
-    ev.sum(slots)
-
   //   [3,5,6,1]
   // [1,2,4,3]
   //   [5,9,9,1]
+  /**
+   * - Step both forward in time.
+   * - Get an iterator BACK in time from the current index
+   * - `zip` will drop the tail of the longer reverseIterator.
+   * - reverse again before building the new Windows to maintain the
+   *   ring buffer structure, and set the index pointer to the last
+   *   item.
+   */
   def +(rhs: Windows[A])(implicit ev: Monoid[A]): Windows[A] = {
     require(lhs.step == rhs.step)
     val currTime = lhs.time max rhs.time
-    val lhs2 = lhs.step(currTime)
-    val rhs2 = rhs.step(currTime)
-    val bldr = Vector.newBuilder[A]
-
-    (0 until (lhs2.slots.size min rhs2.slots.size)).foreach { i =>
-      val li = makeIndex(lhs2.index + i)
-      val ri = makeIndex(rhs2.index + i)
-      val lx = lhs2.slots(li)
-      val rx = rhs2.slots(ri)
-      val zx = ev.plus(lx, rx)
-      bldr += zx
-    }
-    Windows(bldr.result(), 0, lhs2.step, currTime)
+    val newBuf = lhs.step(currTime).buf + rhs.step(currTime).buf
+    Windows(newBuf, step, currTime)
   }
+
+  def lowerBoundSum(implicit ev: Monoid[A]): A = ev.sum(buf.iterator.drop(1))
```

Comment:
I reworked this to just drop the first thing off of the iterator, vs explicitly mucking with `slots`.